api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto

```diff
@@ -0,0 +1,154 @@
+
+  // The application domain to use when calling the service.
+  string domain = 2;
+
+  // Unified matchers.
+  xds.type.matcher.v3.Matcher bucket_matchers = 3;
+}
+
+// [#next-free-field: 6]
+message RateLimitQuotaBucketSettings {
+  // Configures the behavior when the first request has been matched to
+  // the bucket, and the data plane is waiting for the RLQS to return the
+  // first quota assignment.
+  message NoAssignmentBehavior {
+    oneof behavior {
+      option (validate.required) = true;
+
+      // Allow or Deny the requests until the bucket is assigned a quota.
+      // If unset, allow all.
+      type.v3.RateLimitStrategy.BlanketRule blanket_rule = 1;
+
+      // Queue the requests until the bucket is assigned a quota.
+      // The data plane maintains a separate queue for each bucket that has
+      // pending requests.
+      // When the server returns quota assignment, the data plane begins
+      // dequeuing the requests in the FIFO order, and processing them according
+      // to the assignment. No individual request/response matching needed.
+      //
+      // .. note::
+      // To avoid memory exhaust, queue configuration allows to
+      // to impose constraints on the maximum number of pending requests
+      // in the queue, and choose what to do with the request over the limit.
+      type.v3.RateLimitStrategyQueue queue_requests = 2;
+    }
+  }
+
+  // Specifies the behavior when the bucket's assignment has expired, and
+  // hasn't been refreshed due to server's unavailability.
+  message ExpiredAssignmentBehavior {
+    message FallbackLastAssignment {
+      // Limit the time the last assignment is used as a fallback.
+      // [#comment:TODO(sergiitk): what to do after that?]
+      google.protobuf.Duration time_to_live = 1;
+    }
+
+    oneof behavior {
+      option (validate.required) = true;
+
+      // Pre-configured local rate limit strategy.
+      type.v3.RateLimitStrategy fallback_rate_limit = 1;
+
+      // Reuse last quota assignment associated to the bucket.
+      FallbackLastAssignment fallback_last_assignment = 2;
+    }
+  }
+
+  message DenyResponseSettings {
+    // Status code to use to fail requests when quota is exceeded.
+    // [#comment:TODO: figure out the right defaults for these fields!]
+    int32 http_status_code = 1;
+    // For HTTP requests.
+
+    google.rpc.Status deny_grpc_status = 2;
+    // For gRPC requests.
+
+    // A response body to send to the downstream client when the response code
+    // is not OK.
+    bytes http_raw_body = 3;
+
+    // A list of headers to add to the response when the request is throttled.
+    repeated config.core.v3.HeaderValue response_headers_to_add = 4;
+  }
+
+  message BucketIdBuilder {
+    message ValueBuilder {
+      oneof value_specifier {
+        string string_value = 1;
+
+        config.core.v3.TypedExtensionConfig custom_value_builder = 2;
+      }
+    }
+
+    map<string, ValueBuilder> bucket_id_builder = 1;
```

Comment:
What is this mapping from?

Reply:
Documentation here is in progress. Briefly: it's be a map from a string key to `HttpCelMatchInput` (via `TypedExtensionConfig`, so other HTTP inputs should work too). Simplified example (by omitting `TypedExtensionConfig`) :

```yaml
'name': 'hello' ,                     # it can be just a string
'env': 'request.headers["env"]`,      # or can be a cel expression
'user': 'request.headers["user_id"]`,
```

The values will be substituted from the request attributes into `BucketId` (which is simple `map<string, string>`), f.e.

```yaml
'env': 'prod',    # resolved from request.headers["env"]
'name': 'hello',  # stays the same
'user': 'bob',    # resolved from request.headers["user_id"]
```

Reply:
Documented in https://github.com/envoyproxy/envoy/pull/19793/commits/4b8dbd1e2afae1d339519bd8aba3b15d66be075e and https://github.com/envoyproxy/envoy/pull/19793/commits/fe39450b0d317afb0f5674e3f45ad723a86cadf3.

Note that I moved `BucketId` to `io.envoyproxy.envoy.extensions.common.ratelimit.v3`. And added an option to _just_ specify the `BucketId` (not the builder) in the filter configuration, `RateLimitQuotaBucketSettings`. I think it makes sense from the consistency perspective - and makes it easier to argue about for documentation and providing examples.

Reply:
After speaking with @markdroth, I'll roll back `BucketId`, and only keep the builder. There's no real functional use for it, and the builder is good enough.